### ISSUE TITLE
Set up a benchmark suite to assess performance and check for improvment or regression

### ISSUE THREAD

**Issue by Remi-Gau** [nilearn/nilearn#4573]
This kind of fix makes me think it'd be good to have a bunch of benchmark suite to assess performance and check for improvment or regression.

@htwangtw I think you had a look into this, no?
- https://github.com/nilearn/nilearn/issues/3313

By the way @htwangtw if you have time for a review of this one: it is a small one.

_Originally posted by @Remi-Gau in https://github.com/nilearn/nilearn/issues/4571#issuecomment-2379663727_

**Comment by man-shu** [comment]
> Yes that cannot hurt.
> 
> But I think we need to start having some performance benchmark running regularly and some setup to help us do some profiling.
> 
> Am afraid that talking about or trying to optimize performance without measuring things may lead to focus on the wrong part of the code or may even makes things slower.
> Also I'd like to have a way to flag a regression in performance the same way we have regression tests.
> 
> Some of those could also be part of our Doc.
> 
> See sklearn benchmark 
> 
> https://scikit-learn.org/scikit-learn-benchmarks/ 
> 
> See fmriprep
> 
> https://fmriprep.org/en/stable/benchmarks.html
> 
> https://fmriprep.org/en/stable/faq.html#how-much-cpu-time-and-ram-should-i-allocate-for-a-typical-fmriprep-run 

 _Originally posted by @Remi-Gau in [#4811](https://github.com/nilearn/nilearn/issues/4811#issuecomment-2543042195)_

**Comment by man-shu** [comment]
I think #5128, #5152, and [doc page](https://nilearn.github.io/dev/auto_examples/07_advanced/plot_mask_large_fmri.html) together have helped me understand our performance issues a bit better and have also uncovered some potential issues and non-issues:
1. [not much wrong with NiftiMasker on single thread](https://output.circle-artifacts.com/output/job/e1be98b5-0e3d-4aab-8777-a55360d18cef/artifacts/0/dev/auto_examples/07_advanced/plot_compare_img_loading_single.html)
2. [running NiftiMasker in parallel across masks is bad
](https://output.circle-artifacts.com/output/job/e1be98b5-0e3d-4aab-8777-a55360d18cef/artifacts/0/dev/auto_examples/07_advanced/plot_compare_img_loading_parallel.html#run-the-main-function-and-plot-memory-usage)
3. [more memory usage by `mean_img` when using `nilearn.image.load_img` vs. `nibabel.load`](https://github.com/nilearn/nilearn/pull/5128#discussion_r1998318678)

So, if we want to optimize things and compare changes or even just analyze all variations of NiftiMasker (say with smoothing, detrending, etc.) more systematically, it is important to have a proper benchmarking suite setup.

I found some more relevant information on sklearn docs: https://scikit-learn.org/dev/developers/contributing.html#monitoring-performance

And ASV docs: https://asv.readthedocs.io/en/latest/using.html

We don't need our contributors to run these benchmarks on their own just yet (as sklearn wanted: https://github.com/scikit-learn/scikit-learn/issues/16723)

> ![Image](https://github.com/user-attachments/assets/147cbcb9-3f9c-420a-aaaa-17159d952cf8)

# TL;DR

I am thinking of making a separate repo to try out ASV benchmarks and see how useful it could be for us. Will post updates here.

**Comment by man-shu** [comment]
I have started a repo here for this: https://github.com/man-shu/nilearn_benchmarks 

the benchmarks are hosted here: https://man-shu.github.io/nilearn_benchmarks

It has some simple benchmarks that we have in our previous examples as of now. It is only being run on nilearn's latest commit and not any previous commits. 

So, right now, the idea is to only compare with a reference (like loading with nibabel.load vs. nilearn.load_img) and not across commits.

**Comment by bthirion** [comment]
Why not https://github.com/nilearn/nilearn_benchmarks ?

**Comment by man-shu** [comment]
It seems that the limit on the number of concurrent Github CI jobs is per organization:

https://github.com/orgs/community/discussions/26343

The benchmarks take a good amount of time (and I will add more), so I did not want to hold up our PRs at least while we are just testing them out.

**Comment by man-shu** [comment]
So, having played around a bit with ASV I think we can use it in three different ways:

1. we can have two kinds of benchmarks: 
        - one, that compares our implementation with a baseline (eg, default NiftiMasker vs. numpy masking) 
        - second, that runs an object with all different input parameters (eg, NiftiMasker with/without smoothing)
2. the second type of benchmarks could be run on the main daily, just to track the performance over time.
3. run specific benchmarks on PRs (similar to our doc build) to compare the benefits/deficits from a given modification.

All this would be more convenient if we had them on our main nilearn repo. That would also allow us to have more eyes on the benchmarks as I add them.

# TL;DR
Move the benchmarks to the main nilearn repo; run daily on main; allow PRs to run specific benchmarks and compare with main

**Comment by Remi-Gau** [comment]
my hunch would be to not run things on PR by default (but maybe have the possibility to maybe trigger it via commit message): we want to have quick feedback on things on PR so I would not add another workflow there by default that could delay everything else

if a PR affects performance we should see it when it's merged in main and identifying the culprit should be fairly easy

for the same reason: having daily runs on main may not be usful. I don't think we merge things daily anyway... although have more time-points on benchmark time-series can give us a better idea of the variance of the benchmark and and figure out when some change introduces by a PR create some outlier value during benchmarking: this may be even more appropriate where we run things in CI where we do not have perfect control of the architecture we run things and this may affect benchmark results too

**Comment by Remi-Gau** [comment]
so TL;DR:
 - let's not do PR just yet
 - comparing implentation (e.g nilearn / nibabel) : yes
 - different parameters : maybe for some obvious parameters that we expect to be commonly used
 - let's run daily on main to establish some baseline get some feel for what results look like
 - add possibility to trigger the workflow manually

**Comment by emdupre** [comment]
> Move the benchmarks to the main nilearn repo; 

Sorry, I was going to comment on the PR but I realized it might be more trackable to revive this point of the discussion : Why was the decision made to move them into the main repo ? I'm not sure that I see the advantages from what's outlined here.

Is there a strong reason to not re-use the [scikit-learn-bench](https://github.com/scikit-learn/scikit-learn-benchmarks) structure ? Note that they _do_ run them daily, even outside of the main repo, but I don't think we need to do that given our relative pace of development.

**Comment by man-shu** [comment]
> Is there a strong reason to not re-use the [scikit-learn-bench](https://github.com/scikit-learn/scikit-learn-benchmarks) structure

Oooh, I did not come across this repo. From the logs, it looks like their last couple of benchmarks didn't run successfully.

Even though I see that before, we have ended up doing more or less the same thing here: https://github.com/nilearn/benchmarks, except it looks like they are running the benchmarks on a self-hosted runner (deducing from [this machine info](https://github.com/scikit-learn/scikit-learn-benchmarks/blob/46767c8da197225e4c61e43e49dc7a22bcfd0b5d/logs/log_9dee432b#L85))

> Why was the decision made to move them into the main repo ?

At this point, I would just like to have more eyes on the code for benchmarks (via PRs).

The second reason is that I also wanted to have the possibility to run comparisons between the main and PR.
But as Remi suggested, we are keeping that for later.

**Comment by Remi-Gau** [comment]
sklearn have their benchmarks in the main repo, no?

https://github.com/scikit-learn/scikit-learn/tree/main/benchmarks

https://github.com/scikit-learn/scikit-learn/tree/main/asv_benchmarks

**Comment by man-shu** [comment]
Yes, but this repo (https://github.com/scikit-learn/scikit-learn-benchmarks) is where they store all the results, logs and deploy the webpage.
It is analogous to our https://github.com/nilearn/benchmarks
